### DISCUSSION TITLE
Disable dark mode on certain dom element

### DISCUSSION THREAD

**Discussion by kaguya3222** [tailwindlabs/tailwindcss#10646] [Ideas]
I think it would be useful to have the ability to disable dark mode on certain elements. 

For example, in Tailwind docs, we can declare darkMode option as following:

```
{ darkMode: 'class' }
```

Then we're enabling dark mode globally, for example:

```
<html class="dark"> ... </html>
```

Then use dark mode classes like:

```
<html class="dark">
  <div class="text-light-100 dark:text-dark-100">Hello!</div>
</html>
```

But I faced a situation when I need to disable dark mode only in one element (by some condition). Something like this:

```
<html class="dark">
  <div class="text-light-100 dark:text-dark-100">Dark text</div>
  <div>
    <div class="text-light-100 dark:text-dark-100">This text shouldn't be dark if some condition is true</div>
  </div>
</html>
```

Is it possible to have some API that will allow to disable dark mode on some node element ? By applying `light` CSS class for example:

```
<html class="dark">
  <div class="text-light-100 dark:text-dark-100">Dark text</div>
  <div class="light">
    <div class="text-light-100 dark:text-dark-100">This text is light</div>
  </div>
</html>
```

**Comment by kaguya3222** [comment]
@adamwathan Could you please check if this is possible & needed? If this API was approved I would work on this

**Comment by matyushen** [comment]
Any good solutiuon for this yet?

**Reply by kaguya3222** [reply]
I didn't find any good solution using Tailwind. I decided to use javascript self-written solution in 1 component where I needed this behavior. Unfortunately 😢

**Reply by sedubois** [reply]
@kaguya3222 would you mind sharing the outline of your JS solution? It might be interesting as a workaround for https://github.com/tailwindlabs/tailwindcss-typography/issues/228#issuecomment-1561370773. Thanks 🙏

**Reply by kaguya3222** [reply]
@sedubois not sure whether it helps, but I will try to find the code and send it as a gist :) Since it's a React component. 

I would be really happy if some maintainer of tailwind approved this suggestion and I could work on it though

**Comment by renatorib** [comment]
Just sharing my workaround right know:

```js
plugins: [
  plugin(({ addVariant }) => {
    addVariant("dark", [":is(.dark &):not(.light &)"]);
  })
]
```

```html
<div class="dark">
  It's dark!
  <div class="light">
    It's back to light!
  </div>
</div>
```

It changes dark variant to `:is(.dark &):not(.light &)` instead of just `:is(.dark &)`. 
So everything inside `.dark` is still dark mode, unless inside `.light`.

The issue with this approach is that you cannot back to dark mode inside .light anymore.
```html
<div class="dark">
  It's dark!
  <div class="light">
    It's back to light!
    <div class="dark">
      It's still light :(
    </div>
  </div>
</div>
```

![image](https://github.com/tailwindlabs/tailwindcss/assets/3277185/c917d487-ef0d-4806-ba7a-81d88277c95e)

**Reply by VinceBT-BG** [reply]
@renatorib Can you share your Tailwind version and whole config ? How did you make this work, I cannot reproduce on my machine

**Reply by cwqt** [reply]
As far as I can tell trying to override `dark` doesn't work, see https://play.tailwindcss.com/OGTsamETju

You can get around this however by doing:

```ts
plugins: [
  plugin(({ addVariant }) => {
    addVariant("_dark", [":is(.dark &):not(.light &)"]);
  })
]
```

```html
<div class="dark">
  It's dark!
  <div class="light">
    <div class="text-[#00ff00] _dark:text-[#ff0000]">
      It's back to light!
    </div>
  </div>
</div>
```

**Reply by olets** [reply]
> As far as I can tell trying to override dark doesn't work

@cwqt https://github.com/tailwindlabs/tailwindcss/discussions/11534#discussioncomment-8027194

**Reply by rozsazoltan** [reply]
Related:
* https://github.com/tailwindlabs/tailwindcss/discussions/10646#discussioncomment-14193029

```css
@custom-variant dark (&:is(.dark &):not(.light &)); /* just transferred original example from v3 to v4 syntax */
```

* https://github.com/tailwindlabs/tailwindcss/discussions/10646#discussioncomment-14192969 (**recommended**)

```css
@custom-variant dark (&:where(.dark, .dark *:not(.light))); /* recommended, read reference for more details */
```

**Comment by johannbuscail** [comment]
Any update ?

**Comment by olets** [comment]
Here are two solutions, one using [arbitrary variants](https://tailwindcss.com/docs/adding-custom-styles#arbitrary-variants) and one using a [static variant plugin](https://tailwindcss.com/docs/plugins#static-utilities).

https://play.tailwindcss.com/mfu56JScaE

**Comment by VinceBT-BG** [comment]
Try `darkMode: ['variant', [':is(.dark &):not(.light &)']]` in your Tailwind config

**Reply by johannbuscail** [reply]
Doesn't work for me.

**Reply by VinceBT-BG** [reply]
@johannbuscail Are you using `3.4.1` ?

**Reply by johannbuscail** [reply]
Yes, nothing happens.

**Reply by VinceBT-BG** [reply]
Make sure everything is like this:

- Using Tailwind `3.4.1` or higher
- Having this kind of Tailwind config:
```
module.exports = {
	darkMode: ['variant', [':is(.dark &):not(.light &)']],
	// Rest of the config here
}
```
- Having your top `<html>` tag with a theme class for example <html class="dark">
- Having a child of this html tag have the opposite theme class for exemple here <div class="light">

So, this kind of configuration:

```
<html class="dark">
  <body>
  	<div class="light">
		<p class="text-red-500 dark:text-green-500">Hello World!</p>
	</div>
  </body>
</html>
```

The paragraph should be red even though your html tag has the `dark` class

**Reply by Boston343** [reply]
I'm not sure why, but it didn't quite work for me. But THIS does `darkMode: ["variant", "&:is(.dark *):not(.light *)"],`

**Reply by aminkhadivar** [reply]
> I'm not sure why, but it didn't quite work for me. But THIS does `darkMode: ["variant", "&:is(.dark *):not(.light *)"],`

thanks a lot . this is working for me

**Reply by rozsazoltan** [reply]
For v4:

* https://play.tailwindcss.com/0w2UpG3rE1 (I tweaked the playground a bit - it's green if we get the expected result, red if not.)

```css
@custom-variant dark (&:is(.dark &):not(.light &));
```

* https://play.tailwindcss.com/pwHXBAYqr6 (I tweaked the playground a bit - it's green if we get the expected result, red if not.)

```css
@custom-variant dark (&:is(.dark *):not(.light *));
```

Note: I just transferred the original v3 example into v4. It could use some fine-tuning. Currently, unfortunately, the dark mode only works on the first `.dark`, and then `.light` permanently overrides it in the nested elements.

But with this, inside `.light` there's no way to switch back to `.dark`.
* See here: https://github.com/tailwindlabs/tailwindcss/discussions/10646#discussioncomment-14192969

For more examples, read the comments below as well. Perhaps this concept fits best with the goal you want to achieve:
* https://github.com/tailwindlabs/tailwindcss/discussions/10646#discussioncomment-14196210

**Comment by rossinek** [comment]
I had a slightly different situation: instead of forcing a light mode, I wanted to **invert the theme**. The component should be light in dark mode and dark in light mode. Maybe someone will find my solution useful:
```js
module.exports = {
  darkMode: ['variant', ':is(.dark &, .dark-invert &):not(.dark .dark-invert &)'],
  // ...
}
```

Then usage:
```html
<div class="dark-invert">
  <p class="text-red-500 dark:text-blue-500">
    I'm red in dark mode and blue in light mode
  </p>
</div>
```

I'd love to see other/better solutions 👀

**Reply by rozsazoltan** [reply]
For v4:

* https://play.tailwindcss.com/nzMTTNpuvr (I tweaked the playground a bit - it's green if we get the expected result, red if not.)

```css
@custom-variant dark (&:is(.dark &, .dark-invert &):not(.dark .dark-invert &));
```

Note: I just transferred the original v3 example into v4. It could use some fine-tuning. Currently, unfortunately, the dark mode only works on the first `.dark`, and then `.dark-invert` permanently overrides it in the nested elements.

But with this, inside `.dark-invert` there's no way to switch back to `.dark`.
* See here: https://github.com/tailwindlabs/tailwindcss/discussions/10646#discussioncomment-14192969

For more examples, read the comments below as well. Perhaps this concept fits best with the goal you want to achieve:
* https://github.com/tailwindlabs/tailwindcss/discussions/10646#discussioncomment-14196210

**Comment by IranMine123** [comment]
hello I know its been since a while for this issue but any solutions for v4?
i tried custom-variant but didn't work

**Reply by rozsazoltan** [reply]
Up.: I tweaked the playground a bit - it's green if we get the expected result, red if not.

Maybe:
* https://play.tailwindcss.com/81KUWdI9Lu

```css
@custom-variant dark {
  /* default dark */
  &:where(.dark, .dark *):not(.dark .light, .dark .light *) {
    @slot;
  }
  /* dark apply inside light */
  &:where(.dark .light .dark, .dark .light .dark *) {
    @slot;
  }
};
```

If inside `.light` you don't want to recursively re-enable dark mode with another `.dark`, then this is sufficient - but in that case, dark mode will remain permanently disabled within `.light`.

* https://play.tailwindcss.com/ZI3FVfaokO?file=css

```css
@custom-variant dark (&:where(.dark, .dark *):not(.dark .light, .dark .light *));
```

For more examples, read the comments below as well. Perhaps this concept fits best with the goal you want to achieve:
* https://github.com/tailwindlabs/tailwindcss/discussions/10646#discussioncomment-14196210

**Reply by IranMine123** [reply]
> ```css
> @custom-variant dark {
>   /* default dark */
>   &:where(.dark, .dark *):not(.dark .not-dark, .dark .not-dark *) {
>     @slot;
>   }
>   /* dark apply inside not-dark */
>   &:where(.dark .not-dark .dark, .dark .not-dark .dark *) {
>     @slot;
>   }
> };
> ```

thanks for the reply, but the selector you used will broke if there are more than one nesting elements with `light` class.
the solution i came up with (but not a 100% solution still) is:

```css
@custom-variant dark {
  &:is(.dark, .dark *:not(.light, .light *)) {
    @slot;
  }
}
```

**Reply by rozsazoltan** [reply]
I tweaked the playground a bit - it's green if we get the expected result, red if not.

* https://play.tailwindcss.com/Wcv2gPy3D2?file=css (example for your selector)

Yeah, I played around with it, got just as far as you did, so I didn't share it - since there's not much point if I can't reapply the dark * afterwards.

Anyway, the most I can imagine is `.dark .light .dark` - I think it's already a far-fetched example if you need to switch themes more than once. At that point, I'd say it's more of a design flaw, so I was just chasing the impossible purely for fun.

**Reply by rozsazoltan** [reply]
In some cases, it might be acceptable that for every element where we don't want dark mode to apply, we need to use `.light`, and `.light` does not override `.dark` on nested elements.

* https://play.tailwindcss.com/O9bri481sn?file=css

```css
@custom-variant dark (&:where(.dark, .dark *:not(.light)));
```
